### PR TITLE
Update waypoint scoring

### PR DESCRIPTION
**Pull request by yogishadixit**
* Score waypoints using formula `max(0, (100ft -distance) / 100ft)`
* Report scores for best sequence
* Report closest approach to waypoints
* Update tests

### CODE REVIEW

File: server/auvsi_suas/models/mission_config_test.py [hunk 1]
```diff
@@ -107,38 +107,33 @@ def test_satisfied_waypoints(self):
...
         def assertSatisfiedWaypoints(expect, got):
             """Assert two satisfied_waypoints return values are equal."""
-            self.assertEqual(expect[0], got[0])
-            self.assertEqual(expect[1], got[1])
-            for k in expect[2].keys():
-                self.assertIn(k, got[2])
-                self.assertAlmostEqual(expect[2][k], got[2][k], delta=0.1)
-            for k in got[2].keys():
-                self.assertIn(k, expect[2])
+            assertDictionaryEqual(expect[0], got[0])
```
Comment (pmtischler) [inline server/auvsi_suas/models/mission_config_test.py]: Why did you add a custom assert? Is it to have a threshold for floats?

File: server/auvsi_suas/models/mission_config.py [hunk 2]
```diff
@@ -121,16 +121,18 @@ def satisfied_waypoints(self, uas_telemetry_logs):
 
         waypoints = self.mission_waypoints.order_by('order')
 
-        best = {"satisfied": 0, "satisfied_track": 0}
-        current = {"satisfied": 0, "satisfied_track": 0}
+        best = {i: 0.0 for i in range(0, len(waypoints))}
```
Comment (pmtischler) [inline server/auvsi_suas/models/mission_config.py]: This function had a bad assumption that we need to fix now that track distance has been removed. The assumption was that quads would never go backwards through waypoints (likely true), and that planes might want to but couldn't without violating track distance (likely true). Consider what happens now if a plane decides to attempt flying past a waypoint multiple times to attempt a higher score, but happens to touch the next waypiont while turning around. In the past, such would violate the track boundary for sure, but now it'd be valid but mis-scored. The greedy approach here probably won't be able to account for such. You probably need to do a dynamic programming approach using all points within waypoint threshold. That being said, I'd prefer to submit this first and send a follow-up PR with a test case for this edge case and the new implementation.

For each time-sorted point, the dynamic algorithm could be:

for each waypoint:
  for each data point:
    if first waypoint:
      set value of v[waypoint, datapoint] to score_waypoint()
      set value of sv[waypoint, datapoint] to the same
      continue
    get previous waypoint value as max(prev waypoint, prev datapoint)
    set value of v[waypoint, datapoint] to 
          max(score_waypoint(cur waypoint, cur datapoint) + v[prev waypoint, prev datapoint],
                  v[cur waypoint, prev datapoint])
return max over all of d, trace back to get closest on best path

For N waypoints and M data points, where M >> N, runtime would be O(NM).

You can further optimize by only tracking/computing/iterating over data points relevant for a waypoint (ones under threshold).

Comment (pmtischler) [inline server/auvsi_suas/models/mission_config.py]: Updated the prev comment removing some cruft from the psuedocode.

File: server/auvsi_suas/models/mission_config.py [hunk 3]
```diff
@@ -271,11 +253,10 @@ def evaluate_teams(self, users=None):
...
 
             # Determine if the uas hit the waypoints.
-            waypoints_hit, waypoints_hit_track, waypoints_closest = \
+            waypoint_scores, closest_approaches = \
                 self.satisfied_waypoints(uas_logs)
-            eval_data['waypoints_satisfied'] = waypoints_hit
-            eval_data['waypoints_satisfied_track'] = waypoints_hit_track
-            eval_data['waypoints_closest'] = waypoints_closest
+            eval_data['waypoint_scores'] = waypoint_scores
+            eval_data['waypoint_closest_approaches'] = closest_approaches
```
Comment (pmtischler) [inline server/auvsi_suas/models/mission_config.py]: That is, we want to report they closest they got over all time, and the closest they got on the highest scoring sequence. To prevent cross-system scoring inconsistency, we want to keep waypoint_scores.

**Comment by pmtischler** [comment]
Don't forget to squash before submit. Also, use "For #x" in the description so it gets referenced in the issue. When the PR fixes it, you can use Fixes instead of For and it will auto close the issue for you. This PR doesn't fix yet (see comments).

Let's also have @prattmic review.
 
This PR LGTM for merging. Left note about a follow-up PR to handle a new-ish edge case.

File: server/auvsi_suas/models/mission_config_test.py (hunk 1 above)
Comment (yogishadixit) [inline server/auvsi_suas/models/mission_config_test.py]: Yes. The provided assertDictEquals doesn't do float comparison well.

File: server/auvsi_suas/models/mission_config_test.py [hunk 4]
```diff
@@ -107,38 +107,33 @@ def test_satisfied_waypoints(self):
         user = User.objects.create_user('testuser', 'testemail@x.com',
                                         'testpass')
 
+        def assertDictionaryEqual(expect, got):
```
Comment (prattmic) [inline server/auvsi_suas/models/mission_config_test.py]: `def assertDictionaryEqual(self, expect, got):`

I'm not sure why assertSatisifiedWaypoints doesn't pass self either.

File: server/auvsi_suas/models/mission_config_test.py [hunk 5]
```diff
@@ -107,38 +107,33 @@ def test_satisfied_waypoints(self):
...
             """Assert two satisfied_waypoints return values are equal."""
-            self.assertEqual(expect[0], got[0])
-            self.assertEqual(expect[1], got[1])
-            for k in expect[2].keys():
-                self.assertIn(k, got[2])
-                self.assertAlmostEqual(expect[2][k], got[2][k], delta=0.1)
-            for k in got[2].keys():
-                self.assertIn(k, expect[2])
+            assertDictionaryEqual(expect[0], got[0])
+            assertDictionaryEqual(expect[1], got[1])
```
Comment (prattmic) [inline server/auvsi_suas/models/mission_config_test.py]: `self.assertDictionaryEqual(expect[1], got[1])`

File: server/auvsi_suas/models/mission_config.py [hunk 6]
```diff
@@ -121,16 +121,18 @@ def satisfied_waypoints(self, uas_telemetry_logs):
```
Comment (prattmic) [inline server/auvsi_suas/models/mission_config.py]: The docstring above needs to be updated.

(Why doesn't Github let me leave comments anywhere in the file??)

**Comment by yogishadixit** [comment]
Added `waypoints_closest_for_scores`, which changed code structure a bit. PTAL, thanks!

File: server/auvsi_suas/models/mission_config.py [hunk 7]
```diff
@@ -106,10 +106,10 @@ def satisfied_waypoints(self, uas_telemetry_logs):
         Args:
             uas_telemetry_logs: A list of UAS Telemetry logs.
         Returns:
-            satisifed: Total number of waypoints hit (must be in order).
-            satisfied_track: Total number of waypoints hit while remaining
-                along the track. Once the UAV leaves the track no
-                more waypoints will count towards this metric.
+            best: Dictionary of distances to waypoints of the highest scoring
```
Comment (pmtischler) [inline server/auvsi_suas/models/mission_config.py]: distance

**Comment by yogishadixit** [comment]
Fixes #174.
